File: python/stack_queue/truck.py
# 다리를 지나는 트럭(https://programmers.co.kr/learn/courses/30/parts/12081)
# 입출력 예시
# bridge_length	weight	truck_weights	                return
# 2	            10	    [7,4,5,6]	                    8
# 100	        100	    [10]	                        101
# 100	        100	    [10,10,10,10,10,10,10,10,10,10]	110

def solution(bridge_length, weight, truck_weights):
    answer = 0
    bridge=[]
    time=[]
    while len(truck_weights)!=0:
        if len(time)!=0 and answer-time[0]==bridge_length:
            bridge.pop(0)
            time.pop(0)
        if truck_weights[0]+sum(bridge)<=weight and len(bridge)<=bridge_length:
            bridge += [truck_weights.pop(0)]
            time += [answer]
        # print(bridge,time)
        answer += 1
    answer = time[-1]+bridge_length+1
    return answer

File: javascript/stack_queue/truck.js
function solution(bridge_length, weight, truck_weights) {
    var answer = 0;
    let bridgeWeight = 0;
    let index = 0;
    let time = []
    let w = []
    while (index < truck_weights.length) {
        answer += 1
        if (answer - time[0] === bridge_length) {
            bridgeWeight -= w[0]
            time.shift()
            w.shift()
        }
        if (bridgeWeight + truck_weights[index] <= weight) {
            bridgeWeight += truck_weights[index]
            w.push(truck_weights[index])
            time.push(answer)
            index += 1
        }
    }
    return answer + bridge_length;
}

console.log(solution(2, 10, [7,4,5,6]))
console.log(solution(100, 100, [10]))
console.log(solution(100, 100, [10,10,10,10,10,10,10,10,10,10]))